### ISSUE TITLE
Project

### ISSUE THREAD

**Issue by parsaanooshfar-lab** [parsaanooshfar-lab/homework-machine-learning#34]
---

ساختار پروژه

handball_talent_ai/
│
├── [app.py](https://search.eitaa.com/?url=https://app.py)                 # UI (Streamlit)
├── text_[model.py](https://search.eitaa.com/?url=https://model.py)          # استعداد‌یابی متنی
├── pose_[analysis.py](https://search.eitaa.com/?url=https://analysis.py)       # Pose Detection
├── match_[prediction.py](https://search.eitaa.com/?url=https://prediction.py)    # پیش‌بینی مسابقه
├── data/
│   └── handball_[players.csv](https://search.eitaa.com/?url=https://players.csv)
├── [requirements.txt](https://search.eitaa.com/?url=https://requirements.txt)

---

[requirements.txt](https://search.eitaa.com/?url=https://requirements.txt)

streamlit
pandas
numpy
scikit-learn
opencv-python
mediapipe

---

دیتاست (data/handball_[players.csv](https://search.eitaa.com/?url=https://players.csv))

نمونه واقعی‌نما:

age,height,weight,experience,position,label
18,182,78,3,back,1
20,176,70,5,wing,1
22,190,90,6,pivot,1
19,170,68,2,wing,1
24,185,88,8,goalkeeper,1
21,168,72,1,none,0

label = مناسب هندبال (1 بله، 0 نه)

---

text_[model.py](https://search.eitaa.com/?url=https://model.py) (مرحله اول)

import pandas as pd
from [sklearn.tree](https://search.eitaa.com/?url=https://sklearn.tree) import DecisionTreeClassifier

def train_text_model():
    df = [pd.read](https://search.eitaa.com/?url=https://pd.read)_csv("data/handball_[players.csv](https://search.eitaa.com/?url=https://players.csv)")
    X = df[["age","height","weight","experience"]]
    y = df["label"]

    model = DecisionTreeClassifier(max_depth=4)
    [model.fit](https://search.eitaa.com/?url=https://model.fit)(X, y)
    return model

def predict_player(model, age, height, weight, experience):
    pred = [model.predict](https://search.eitaa.com/?url=https://model.predict)([[age, height, weight, experience]])
    return "مناسب هندبال" if pred[0] == 1 else "نامناسب"

---

pose_[analysis.py](https://search.eitaa.com/?url=https://analysis.py) (مرحله دوم)

import cv2
import mediapipe as mp
import math

mp_pose = [mp.solutions.pose](https://search.eitaa.com/?url=https://mp.solutions.pose)

def analyze_pose(video_path):
    cap = [cv2.VideoCapture](https://search.eitaa.com/?url=https://cv2.VideoCapture)(video_path)
    pose = mp_[pose.Pose](https://search.eitaa.com/?url=https://pose.Pose)()
    scores = []

    while [cap.isOpened](https://search.eitaa.com/?url=https://cap.isOpened)():
        ret, frame = [cap.read](https://search.eitaa.com/?url=https://cap.read)()
        if not ret:
            break
        img = [cv2.cvtColor](https://search.eitaa.com/?url=https://cv2.cvtColor)(frame, [cv2.COLOR](https://search.eitaa.com/?url=https://cv2.COLOR)_BGR2RGB)
        result = [pose.process](https://search.eitaa.com/?url=https://pose.process)(img)
        if [result.pose](https://search.eitaa.com/?url=https://result.pose)_landmarks:
            shoulder = [result.pose](https://search.eitaa.com/?url=https://result.pose)_[landmarks.landmark](https://search.eitaa.com/?url=https://landmarks.landmark)[mp_[pose.PoseLandmark.RIGHT](https://search.eitaa.com/?url=https://pose.PoseLandmark.RIGHT)_SHOULDER]
            elbow = [result.pose](https://search.eitaa.com/?url=https://result.pose)_[landmarks.landmark](https://search.eitaa.com/?url=https://landmarks.landmark)[mp_[pose.PoseLandmark.RIGHT](https://search.eitaa.com/?url=https://pose.PoseLandmark.RIGHT)_ELBOW]
            wrist = [result.pose](https://search.eitaa.com/?url=https://result.pose)_[landmarks.landmark](https://search.eitaa.com/?url=https://landmarks.landmark)[mp_[pose.PoseLandmark.RIGHT](https://search.eitaa.com/?url=https://pose.PoseLandmark.RIGHT)_WRIST]
            angle = abs(shoulder.y - wrist.y)
            [scores.append](https://search.eitaa.com/?url=https://scores.append)(angle)

    [cap.release](https://search.eitaa.com/?url=https://cap.release)()
    return sum(scores)/len(scores) if scores else 0

---

match_[prediction.py](https://search.eitaa.com/?url=https://prediction.py) (مرحله نهایی)

from [sklearn.linear](https://search.eitaa.com/?url=https://sklearn.linear)_model import LogisticRegression
import numpy as np

def train_match_model():
    X = [np.array](https://search.eitaa.com/?url=https://np.array)([
        [30, 0.6, 10],
        [28, 0.55, 12],
        [35, 0.7, 8],
        [25, 0.5, 14]
    ])
    y = [np.array](https://search.eitaa.com/?url=https://np.array)([1, 1, 1, 0])
    model = LogisticRegression()
    [model.fit](https://search.eitaa.com/?url=https://model.fit)(X, y)
    return model

def predict_match(model, goals, accuracy, fouls):
    prob = [model.predict](https://search.eitaa.com/?url=https://model.predict)_proba([[goals, accuracy, fouls]])[0][1]
    return prob

---

[app.py](https://search.eitaa.com/?url=https://app.py) (UI کامل)

import streamlit as st
from text_model import train_text_model, predict_player
from pose_analysis import analyze_pose
from match_prediction import train_match_model, predict_match
import tempfile

[st.set](https://search.eitaa.com/?url=https://st.set)_page_config(page_title="Handball Talent AI")

[st.title](https://search.eitaa.com/?url=https://st.title)("سیستم هوشمند استعداد‌یابی هندبال")

text_model = train_text_model()
match_model = train_match_model()

[st.header](https://search.eitaa.com/?url=https://st.header)("مرحله اول: تحلیل اطلاعات متنی")
age = [st.number](https://search.eitaa.com/?url=https://st.number)_input("سن", 10, 40)
height = [st.number](https://search.eitaa.com/?url=https://st.number)_input("قد (cm)", 150, 210)
weight = [st.number](https://search.eitaa.com/?url=https://st.number)_input("وزن (kg)", 40, 130)
exp = [st.number](https://search.eitaa.com/?url=https://st.number)_input("سابقه ورزشی (سال)", 0, 20)

if [st.button](https://search.eitaa.com/?url=https://st.button)("تحلیل استعداد"):
    res = predict_player(text_model, age, height, weight, exp)
    [st.success](https://search.eitaa.com/?url=https://st.success)(res)

[st.header](https://search.eitaa.com/?url=https://st.header)("مرحله دوم: تحلیل ویدئو (Pose Detection)")
video = [st.file](https://search.eitaa.com/?url=https://st.file)_uploader("آپلود ویدئو", type=["mp4"])
if video:
    tfile = [tempfile.NamedTemporaryFile](https://search.eitaa.com/?url=https://tempfile.NamedTemporaryFile)(delete=False)
    [tfile.write](https://search.eitaa.com/?url=https://tfile.write)([video.read](https://search.eitaa.com/?url=https://video.read)())
    score = analyze_pose([tfile.name](https://search.eitaa.com/?url=https://tfile.name))
    [st.info](https://search.eitaa.com/?url=https://st.info)(f"امتیاز تکنیکی پرتاب: {round(score,3)}")

[st.header](https://search.eitaa.com/?url=https://st.header)("مرحله نهایی: پیش‌بینی نتیجه مسابقه")
goals = [st.number](https://search.eitaa.com/?url=https://st.number)_input("میانگین گل", 0, 50)
acc = [st.slider](https://search.eitaa.com/?url=https://st.slider)("دقت پرتاب", 0.0, 1.0)
fouls = [st.number](https://search.eitaa.com/?url=https://st.number)_input("تعداد خطا", 0, 30)
if [st.button](https://search.eitaa.com/?url=https://st.button)("پیش‌بینی"):
    p = predict_match(match_model, goals, acc, fouls)
    [st.write](https://search.eitaa.com/?url=https://st.write)(f"احتمال برد: {round(p*100,2)}٪")

---

اجرا

pip install -r [requirements.txt](https://search.eitaa.com/?url=https://requirements.txt)
streamlit run [app.py](https://search.eitaa.com/?url=https://app.py)

---
